File: 09.Greed贪心/leetcode_2033_minOperations.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <unordered_map>
using namespace std;

class Solution {
public:
    int minOperations(vector<vector<int>>& grid, int x) {
        int m = grid.size(), n = grid[0].size();
        unordered_map<int, int> cnt;
        vector<int> nums;
        for (int i=0; i<m; i++){
            for (int j=0; j<n; j++){
                int a = grid[i][j];
                nums.push_back(a);
                cnt[a%x] = 1;
                if (cnt.size() == 2){
                    return -1;
                }
            }
        }
        sort(nums.begin(), nums.end());
        int target = nums[nums.size()/2];
        int res = 0;
        for (int i : nums){
            res += abs(i-target);
        }
        return res/x;
    }
};

int main(){
    Solution s;
    vector<vector<int>> grid = {{2,4},{6,8}};
    int x = 2;
    cout << s.minOperations(grid, x) << endl;
}

File: 09.Greed贪心/leetcode_2033_minOperations.py
from typing import List
from collections import defaultdict

class Solution:
    def minOperations(self, grid: List[List[int]], x: int) -> int:
        cnt = defaultdict(int)
        lst = []
        m, n = len(grid), len(grid[0])
        for i in range(m):
            for j in range(n):
                a = grid[i][j]
                lst.append(a)
                cnt[a%x] = 1
                if len(cnt) == 2:
                    return -1
        lst.sort()
        # 贪心，变成中间大小的一定花费更少
        target = lst[len(lst)//2]
        return sum(abs(target-a) for a in lst)//x

if __name__ == "__main__":
    s = Solution()
    grid = [[2,4],[6,8]]
    x = 2
    print(s.minOperations(grid, x))
